### ISSUE TITLE
Client Side Serialization Compiler Error: "Serialize: Type not supported"

### ISSUE THREAD

**Issue by ascjones** [intellifactory/websharper#468]
Following the basic example here: http://websharper.com/blog-entry/4419/websharper-3-3-released-with-client-side-json-serialization using WebSharper.UI.Next version 3.4.19.168.

Strangely when building via Visual Studio it works okay, however when compiling via MSBuild it fails with the following error:

```
 (WebSharperCompile target) ->
         C:\code\WebSharper.Playground\packages\WebSharper\build\WebSharper.targets(57,5): error : Appli
       cation1.Article: Serialize: Type not supported: Application1.Article [C:\code\WebSharper.Playgrou
       nd\Application1\Application1.fsproj
```

Looking at the [code](https://github.com/intellifactory/websharper/blob/2380a8509b971acef63c8359fcb7ef1ce862cdc1/src/sitelets/WebSharper.Web/ClientSideJson.fs#L416) I can't understand why the IsUnion and IsRecord checks would both fail for the type when being run by MSBuild.

You can get the code to reproduce the issue here: https://github.com/ascjones/WebSharper.ClientSideJsonIssue

Just run build.bat and you should get the error (unless the issue is with my dev environment of course)

(edit: also [tried using nuget](https://github.com/ascjones/WebSharper.ClientSideJsonIssue/tree/nuget) instead of paket just to rule that out. Same error)

**Comment by granicz** [comment]
On a first glance, we compared the command line options passed in both scenarios, and they are indeed different.  But when making them the same by setting `OtherFlags` to the diff, `msbuild` still yielded the same error.  So this needs more investigation.  It is probably something related to how the WebSharper build task gets instantiated, so will look into that when we get a chance.

**Comment by ascjones** [comment]
Thanks for looking into this Adam.

**Comment by ascjones** [comment]
I've looked further at this, and it appears it is to do with FSharp.Core. My project is currently targeting F# 3.1 (FSharp.Core 4.3.1), which succeeds in VS but fails with MSBuild because of the `` Serialize: Type not supported`` error. Debugging shows that ``FSharpType.IsRecord`` is returning false for the type (should be true).

However when I change my project to target F# 3.0 (FSharp.Core 4.3.0.0) then the build succeeds with both VS and MSBuild.

It appears that MSBuild loads the version of FSharp.Core (4.3.0.0) in the same directory as the WebSharper.MSBuild.dll (``packages\WebSharper\tools\net40``), whilst Visual Studio does not (probably loading from the GAC?). I've proved this by deleting the FSharp.Core assemblies from ``packages\WebSharper\tools\net40`` directory and running both builds. MSBuild fails with an assembly load error and VS succeeds!

I feel like I'm nearly there, but have hit a brick wall. Clearly there's some way to get this to work (my project targeting F# 3.1) since it works okay in Visual Studio. Seems that when MSBuild invokes the WebSharperCompile target it should be forced to load 4.3.1.0 instead of 4.3.0.0, as per the binding redirect in my app.config:

```
<dependentAssembly>
    <assemblyIdentity name="FSharp.Core" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
    <bindingRedirect oldVersion="0.0.0.0-4.3.1.0" newVersion="4.3.1.0" />
  </dependentAssembly>
```

**Comment by cgravill** [comment]
Thanks for providing this workaround Andrew. I was able to use it to deal with a related problem #504

In case other people need it, I actually had a later version of msbuild i.e. ```C:\Program Files (x86)\MSBuild\14.0\Bin\MSBuild.exe.config``` and I also needed to ensure there was GAC'd version of F# 4.3.1.0 / F# 3.1 on the relevant build machine.

**Comment by Tarmil** [comment]
@ascjones Can you try updating to the latest WebSharper 3.6.x and removing your `MSBuild.exe.config` workaround? Your issue _should_ be fixed by 92a77c701f421a9dd5a15a15f33992ac53bf5ce1.

**Comment by ascjones** [comment]
Hi @Tarmil, it works!

Thanks you very much indeed :-)
